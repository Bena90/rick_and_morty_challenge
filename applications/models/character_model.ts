import { Character } from "@/domain/entities/character";
import { CharacterDTO } from "../dto/character_dto";

export class CharacterModel implements Character {
  id: number;
  name: string;
  status: string;
  species: string;
  type: string;
  gender: string;
  image: string;
  episode: string[];
  url: string;

  constructor(
    id: number,
    name: string,
    status: string,
    species: string,
    type: string,
    gender: string,
    image: string,
    episode: string[],
    url: string,
  ) {
    this.id = id;
    this.name = name;
    this.status = status;
    this.species = species;
    this.type = type;
    this.gender = gender;
    this.image = image;
    this.episode = episode;
    this.url =   url;
  }

  static fromJson(json: {[key: string]: never}): CharacterModel {
    return new CharacterModel(
      json[CharacterDTO.id] ?? 0,
      json[CharacterDTO.name] ?? 'N/A',
      json[CharacterDTO.status] ?? 'N/A',
      json[CharacterDTO.species] ?? 'N/A',
      json[CharacterDTO.type] ?? 'N/A',
      json[CharacterDTO.gender] ?? 'N/A',
      json[CharacterDTO.image] ?? 'N/A',
      json[CharacterDTO.episode] ?? [],
      json[CharacterDTO.url] ?? 'N/A',
    );
  }
}