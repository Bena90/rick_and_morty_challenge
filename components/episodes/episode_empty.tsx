export function EpisodeEmpty({text}: { text: string }) {
  return (
    <div className="flex justify-center items-center w-full h-[200px]">
      <span className="font-thin">
        {text}
      </span>
    </div>
  );
}