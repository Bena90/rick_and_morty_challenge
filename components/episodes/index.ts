export { CompareList } from './compare_list';
export { EpisodeEmpty } from './episode_empty';
export { EpisodeItem } from './episode_item';
export { EpisodesGrid } from './episodes_grid';
export { EpisodesList } from './episodes_list';
export { ListLayout } from './list_layout';

