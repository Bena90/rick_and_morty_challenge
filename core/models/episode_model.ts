import { EpisodeAdapter } from "../adapters/episode_adapter";
import { Episode } from "../entities/episode";

export class EpisodeModel implements Episode {
  id: number;
  name: string;
  airDate: string;
  episode: string;
  characters: string[];
  url: string;
  created: Date;

  constructor(
    id: number,
    name: string,
    airDate: string,
    episode: string,
    characters: string[],
    url: string,
    created: Date,
  ) {
    this.id = id
    this.name = name
    this.airDate = airDate
    this.episode = episode
    this.characters = characters
    this.url = url
    this.created = created
  }

  static fromJson(json: { [key: string]: never }): EpisodeModel {
    return new EpisodeModel(
      json[EpisodeAdapter.id] ?? '',
      json[EpisodeAdapter.name] ?? '',
      json[EpisodeAdapter.airDate] ?? '',
      json[EpisodeAdapter.episode] ?? '',
      json[EpisodeAdapter.characters] ?? [],
      json[EpisodeAdapter.url] ?? '',
      json[EpisodeAdapter.created] ?? '',
    );
  }
}