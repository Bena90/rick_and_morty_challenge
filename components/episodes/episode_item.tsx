'use client'

import { Episode } from "@/core/entities/episode";
import { useEffect, useState } from "react";
import { EpisodeModel } from '../../core/models/episode_model';

export function EpisodeItem({episode}: {episode: string}) {
  const [ episodeInfo, setEpisodeInfo] = useState<Episode>()

  useEffect(() => {
    const getEpisode = async(episode: string) => {
      if(!episode) return;

      await fetch(episode)
        .then( res => res.json())
        .then((data) => {
          if (data) setEpisodeInfo(EpisodeModel.fromJson(data))
        })
        .catch((error) => console.log(error))
    }

    getEpisode(episode);
  }, [episode])

  if(!episodeInfo) return

  return (
    <div className="flex flex-col bg-slate-900 rounded-xl p-2 border border-slate-800 hover:bg-slate-800  hover:border-slate-700 ">
      <div className="truncate">
        <span className="font-bold tracking-wider text-green">{episodeInfo.name}</span>
      </div>

      <div className="flex gap-4 flex-col lg:flex-row">
        <div className="flex gap-2 flex-col md:flex-row">
          <span className="">Ep.:</span>
          <span className="font-thin"> {episodeInfo.episode}</span>
        </div>
        <div className="flex-col md:flex-row">
          <span className="">Date:</span>
          <span className="font-thin"> {episodeInfo.airDate}</span> 
        </div>
      </div>
    </div>
  );
}