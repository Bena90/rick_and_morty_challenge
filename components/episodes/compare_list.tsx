import { EpisodeEmpty } from './episode_empty';
import { EpisodeItem } from './episode_item';
import { ListLayout } from './list_layout';

export interface Episode {
  id:         number;
  name:       string;
  air_date:   string;
  episode:    string;
  characters: string[];
  url:        string;
  created:    Date;
}

export function CompareList({episodes}: {episodes?: string[]}) {

  return (
    <ListLayout>
      {episodes === undefined
      ? <EpisodeEmpty text={"The selected characters do not share any episode"} />
      : episodes.map((episode) => <EpisodeItem key={episode} url={episode} />)
      }
    </ListLayout>
  );
}

