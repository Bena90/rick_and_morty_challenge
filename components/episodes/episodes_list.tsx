import { EpisodeEmpty } from "./episode_empty";
import { EpisodeItem } from "./episode_item";
import { ListLayout } from "./list_layout";

export interface Episode {
  id:         number;
  name:       string;
  air_date:   string;
  episode:    string;
  characters: string[];
  url:        string;
  created:    Date;
}

export function EpisodesList({
  episodes, characterName
}: {
  episodes?: string[],
  characterName?: string,
}) {

  return (
    <ListLayout>
      <div>
      <span className="text-xl font-light">Character: </span>
      <span className="text-xl text-pink font-black tracking-wider	">
        {characterName ?? 'Not selected'}
      </span>
      </div>
      {episodes
        ? episodes?.map((episode) => <EpisodeItem key={episode} url={episode} />)
        : <EpisodeEmpty text="Please, select one character" />
      }
    </ListLayout>
  );
}

