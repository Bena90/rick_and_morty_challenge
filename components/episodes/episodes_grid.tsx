'use client'
import localFont from 'next/font/local';
import { CompareList } from './compare_list';
import { EpisodesList } from './episodes_list';
import { useEpisodes } from './use_episodes/use_episodes';

const myFont = localFont({ src: '../../core/fonts/rick.ttf' })

export function EpisodesGrid() {
  const { episodesA, episodesB, episodesCompare, characterA, characterB } = useEpisodes();

  return (
    <section className="flex flex-col w-full gap-4">
      <h3 className={`${myFont.className} text-4xl text-green`}>Episodes</h3>
      <h4 className='text-xl text-slate-200 font-thin tracking-wider'><span className='text-pink font-normal'>2 - </span>Now, you can compare which episodes the selected characters share.</h4>
      <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3">
        <EpisodesList episodes={episodesA} characterName={characterA} />
        <CompareList episodes={episodesCompare} />
        <EpisodesList episodes={episodesB} characterName={characterB} />
      </div>
    </section>
  );
}
