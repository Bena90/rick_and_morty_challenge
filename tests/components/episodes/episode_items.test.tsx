import { EpisodeItem, getEpisode } from '@/components/episodes/episode_item';
import { EpisodeModel } from '@/core/models/episode_model';
import { render } from '@testing-library/react';

describe('Test in Episode Item', () => {
  const episode: string = "https://rickandmortyapi.com/api/episode/1"
  
  test('getEpisode should return a valid EpisodeModel object', async() => {
    const response = await getEpisode(episode);

    if (response !== null) {
      expect(response).toBeInstanceOf(EpisodeModel);
      expect(response).toHaveProperty('id');
      expect(response).toHaveProperty('name');
      expect(response).toHaveProperty('airDate');
      expect(response).toHaveProperty('episode');
      expect(response).toHaveProperty('characters');
      expect(response).toHaveProperty('url');
      expect(response).toHaveProperty('created');
    } else {
      fail('Expected response to be an instance of EpisodeModel, but got null');
    }
  });

  test('getEpisode should response undefined if url it`s wrong', async () => {
    const resp = await getEpisode('episode');

    expect(resp).toBeUndefined();
  });

  test('episode item should match with snapshot', () => {
    const {container} = render( <EpisodeItem episode={ episode } /> );

    expect( container ).toMatchSnapshot();
  })

});
