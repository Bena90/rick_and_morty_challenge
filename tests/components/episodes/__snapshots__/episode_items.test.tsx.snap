// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in Episode Item episode item should match with snapshot 1`] = `<div />`;
